pcsx2/USB/usb-pad/usb-realplay.cpp

```diff
@@ -0,0 +1,466 @@
+				ret = usb_desc_handle_control(dev, p, request, value, index, length, data);
+				if (ret >= 0)
+				{
+					return;
+				}
+				p->status = USB_RET_STALL;
+				break;
+		}
+	}
+
+	static void realplay_handle_data(USBDevice* dev, USBPacket* p)
+	{
+		RealPlayState* s = USB_CONTAINER_OF(dev, RealPlayState, dev);
+
+		switch (p->pid)
+		{
+			case USB_TOKEN_IN:
+				if (p->ep->nr == 1)
+				{
+					pxAssert(p->buffer_size >= sizeof(s->data));
+
+					std::memcpy(p->buffer_ptr, &s->data, sizeof(s->data));
+
+					// simulate a slight move to avoid a game "protection" : controller disconnected
+					p->buffer_ptr[0] ^= s->state;
+					s->state ^= 1;
+
+					p->actual_length += sizeof(s->data);
+				}
+				else
+				{
+					goto fail;
+				}
+				break;
+			case USB_TOKEN_OUT:
+				break;
+			default:
+			fail:
+				p->status = USB_RET_STALL;
+				break;
+		}
+	}
+
+	static void realplay_unrealize(USBDevice* dev)
+	{
+		RealPlayState* s = USB_CONTAINER_OF(dev, RealPlayState, dev);
+		delete s;
+	}
+
+	const char* RealPlayDevice::Name() const
+	{
+		return TRANSLATE_NOOP("USB", "RealPlay Device");
+	}
+
+	const char* RealPlayDevice::TypeName() const
+	{
+		return "RealPlay";
+	}
+
+	USBDevice* RealPlayDevice::CreateDevice(SettingsInterface& si, u32 port, u32 type) const
+	{
+		RealPlayState* s = new RealPlayState(port, type);
+
+		s->desc.full = &s->desc_dev;
+		s->desc.str = desc_strings;
+
+		const uint8_t* dev_desc = nullptr;
+		int dev_desc_len = 0;
+
+		switch (s->type)
+		{
+			case REALPLAY_RACING:
+				dev_desc = realplay_racing_dev_descriptor;
+				dev_desc_len = sizeof(realplay_racing_dev_descriptor);
+				break;
+			case REALPLAY_SPHERE:
+				dev_desc = realplay_sphere_dev_descriptor;
+				dev_desc_len = sizeof(realplay_sphere_dev_descriptor);
+				break;
+			case REALPLAY_GOLF:
+				dev_desc = realplay_golf_dev_descriptor;
+				dev_desc_len = sizeof(realplay_golf_dev_descriptor);
+				break;
+			case REALPLAY_POOL:
+				dev_desc = realplay_pool_dev_descriptor;
+				dev_desc_len = sizeof(realplay_pool_dev_descriptor);
+				break;
+			default:
+				pxAssertMsg(false, "Unhandled type");
+				break;
+		}
+
+		if (usb_desc_parse_dev(dev_desc, dev_desc_len, s->desc, s->desc_dev) < 0)
+			goto fail;
+		if (usb_desc_parse_config(config_descriptor, sizeof(config_descriptor), s->desc_dev) < 0)
+			goto fail;
+
+		s->dev.speed = USB_SPEED_FULL;
+		s->dev.klass.handle_attach = usb_desc_attach;
+		s->dev.klass.handle_reset = nullptr;
+		s->dev.klass.handle_control = realplay_handle_control;
+		s->dev.klass.handle_data = realplay_handle_data;
+		s->dev.klass.unrealize = realplay_unrealize;
+		s->dev.klass.usb_desc = &s->desc;
+		s->dev.klass.product_desc = nullptr;
+
+		usb_desc_init(&s->dev);
+		usb_ep_init(&s->dev);
+
+		return &s->dev;
+
+	fail:
+		realplay_unrealize(&s->dev);
+		return nullptr;
+	}
+
+	bool RealPlayDevice::Freeze(USBDevice* dev, StateWrapper& sw) const
+	{
+		RealPlayState* s = USB_CONTAINER_OF(dev, RealPlayState, dev);
+		
```

Comment:
Although not shown in GitHubs diff tool for some reason, there appears to be 2 tabs on this blank line.
If you check [here](https://github.com/PCSX2/pcsx2/raw/76de27392197da01d91257be1e7f93609e711b55/pcsx2/USB/usb-pad/usb-realplay.cpp) it's present.

Reply:
fixed